instructionAPI/src/decoder/riscv/decoder.C

```diff
@@ -0,0 +1,795 @@
+
+      // Determine whether it is bgeu or bleu
+      // It is impossible from telling those two apart without extracting the
+      // raw registers
+      const uint32_t *insn_raw_ptr =
+          reinterpret_cast<const uint32_t *>(insn.ptr());
+      int32_t rs1_enc_raw =
+          ((*insn_raw_ptr) >> REG_RS1_ENC_OFFSET) & REG_ENC_MASK;
+      int32_t rs2_enc_raw =
+          ((*insn_raw_ptr) >> REG_RS2_ENC_OFFSET) & REG_ENC_MASK;
+      int32_t rs1_enc_od = riscv::translate_register(
+                               static_cast<riscv_reg>(operands[0].reg), mode) &
+                           0xff;
+      int32_t rs2_enc_od = riscv::translate_register(
+                               static_cast<riscv_reg>(operands[1].reg), mode) &
+                           0xff;
+      // Compare the raw value of rs1 with operands[0] and rs2 with operands[1]
+      // If they do not match, this means that we're using these 4 pseudo
+      // instructions
+      if (rs1_enc_raw == rs2_enc_od && rs2_enc_raw == rs1_enc_od) {
+        // swap rs1 and rs2
+        res = {operands[1], operands[0], operands[2]};
+        insn.getOperation().updateMnemonic("bgeu");
+      }
+    }
+    break;
+  }
+  default: {
+    break;
+  }
+  }
+  return res;
+}
+
+void riscv_decoder::add_pc_operands(Instruction &insn) {
+  constexpr bool is_read = true;
+  constexpr bool is_write = true;
+  constexpr bool is_implicit = true;
+  const entryID eid = insn.getOperation().getID();
+
+  auto const pc = makeRegisterExpression(MachRegister::getPC(this->m_Arch));
+  switch (eid) {
+  case riscv64_op_auipc: {
+    insn.appendOperand(pc, is_read, !is_write, is_implicit);
+    break;
+  }
+  case riscv64_op_jalr:
+  case riscv64_op_c_jr:
+  case riscv64_op_c_jalr: {
+    insn.appendOperand(pc, !is_read, is_write, is_implicit);
+    break;
+  }
+  case riscv64_op_jal:
+  case riscv64_op_c_j:
+  case riscv64_op_beq:
+  case riscv64_op_bne:
+  case riscv64_op_blt:
+  case riscv64_op_bge:
+  case riscv64_op_bltu:
+  case riscv64_op_bgeu:
+  case riscv64_op_c_beqz:
+  case riscv64_op_c_bnez: {
+    insn.appendOperand(pc, is_read, is_write, is_implicit);
+    break;
+  }
+  default: {
+    break;
+  }
+  }
+}
+} // namespace InstructionAPI
+} // namespace Dyninst
+
+namespace {
+
+std::map<riscv_reg, implicit_state> implicit_registers(cs_detail const *d) {
+  std::map<riscv_reg, implicit_state> regs;
+  for (int i = 0; i < d->regs_read_count; ++i) {
+    regs.emplace(static_cast<riscv_reg>(d->regs_read[i]),
+                 implicit_state{true, false});
+  }
+  for (int i = 0; i < d->regs_write_count; ++i) {
+    auto res = regs.emplace(static_cast<riscv_reg>(d->regs_write[i]),
+                            implicit_state{false, true});
+    if (!res.second) {
+      // Register already existed, so was read. Mark it written.
+      res.first->second.written = true;
+    }
+  }
+  return regs;
+}
+
+cs_riscv_op make_reg_op(int32_t reg, uint8_t access) {
```

Comment:
`op.access` has type enum `1cs_ac_type`. You can't assign an int to enum, so just use the correct type for the parameter here and in the declaration:

```suggestion
cs_riscv_op make_reg_op(int32_t reg, cs_ac_type access) {
```